src/script/cpp_api/s_env.h

```diff
@@ -7,13 +7,55 @@
 #include "cpp_api/s_base.h"
 #include "irr_v3d.h"
 #include "mapnode.h"
+#include "server/blockmodifier.h"
 #include <unordered_set>
 #include <vector>
 
 class ServerEnvironment;
 class MapBlock;
 struct ScriptCallbackState;
 
+/*
+	LuaABM
+*/
+
+class LuaABM : public ActiveBlockModifier {
+private:
+	const int m_id;
+
+	std::string m_name;
+	std::vector<std::string> m_trigger_contents;
+	std::vector<std::string> m_required_neighbors;
+	std::vector<std::string> m_without_neighbors;
+	float m_trigger_interval;
+	u32 m_trigger_chance;
+	bool m_simple_catch_up;
+	s16 m_min_y;
+	s16 m_max_y;
+
+public:
+	LuaABM(int id,
+			const std::string &name,
+			const std::vector<std::string> &trigger_contents,
+			const std::vector<std::string> &required_neighbors,
+			const std::vector<std::string> &without_neighbors,
+			float trigger_interval, u32 trigger_chance, bool simple_catch_up,
+			s16 min_y, s16 max_y);
+
+	virtual const std::string &getName() const;
+	virtual const std::vector<std::string> &getTriggerContents() const;
+	virtual const std::vector<std::string> &getRequiredNeighbors() const;
+	virtual const std::vector<std::string> &getWithoutNeighbors() const;
+	virtual float getTriggerInterval();
+	virtual u32 getTriggerChance();
+	virtual bool getSimpleCatchUp();
+	virtual s16 getMinY();
+	virtual s16 getMaxY();
+
+	virtual void trigger(ServerEnvironment *env, v3s16 p, MapNode n,
+			u32 active_object_count, u32 active_object_count_wider);
+};
```

Comment:
you can forward-declare this instead of moving it

Reply:
I know, but https://github.com/luanti-org/luanti/pull/15992#discussion_r2041204968 . 

If I do forward-declare, I have to use `reinterpret_cast`. Or, is there a way to forward it without needing to use `reinterpret_cast`?

Reply:
@sfan5

Reply:
It is already done:
https://github.com/luanti-org/luanti/blob/ce6e5ed25eec3db7717fbaa97c15aebaa6a3c909/src/script/lua_api/l_env.cpp#L1380-L1392

The problem is still the same. I have to return a class from one file to another, so I have to define it in the header.

I can use the definition `class LuaABM;`, after it, I have to go against https://github.com/luanti-org/luanti/pull/15992#discussion_r2041204968 and use `reinterpret_cast` or move the full class definition to the header file.